OpenRA.Mods.Common/Traits/Radar/AppearsOnRadar.cs

```diff
@@ -18,38 +18,43 @@
 
 namespace OpenRA.Mods.Common.Traits.Radar
 {
-	public class AppearsOnRadarInfo : ITraitInfo
+	public class AppearsOnRadarInfo : ConditionalTraitInfo
 	{
 		public readonly bool UseLocation = false;
 
-		public object Create(ActorInitializer init) { return new AppearsOnRadar(this); }
+		[Desc("Player stances who can view this actor on radar.")]
+		public readonly Stance ValidStances = Stance.Ally | Stance.Neutral | Stance.Enemy;
+
+		public override object Create(ActorInitializer init) { return new AppearsOnRadar(this); }
 	}
 
-	public class AppearsOnRadar : IRadarSignature, INotifyCreated
+	public class AppearsOnRadar : ConditionalTrait<AppearsOnRadarInfo>, IRadarSignature
 	{
-		readonly AppearsOnRadarInfo info;
+		static readonly IEnumerable<Pair<CPos, Color>> NoCells = Enumerable.Empty<Pair<CPos, Color>>();
 		IRadarColorModifier modifier;
 
 		Color currentColor = Color.Transparent;
 		Func<Pair<CPos, SubCell>, Pair<CPos, Color>> cellToSignature;
 
-		public AppearsOnRadar(AppearsOnRadarInfo info)
-		{
-			this.info = info;
-		}
+		public AppearsOnRadar(AppearsOnRadarInfo info) : base(info) { }
 
-		public void Created(Actor self)
+		protected override void Created(Actor self)
 		{
+			base.Created(self);
 			modifier = self.TraitsImplementing<IRadarColorModifier>().FirstOrDefault();
 		}
 
 		public IEnumerable<Pair<CPos, Color>> RadarSignatureCells(Actor self)
 		{
+			var viewer = self.World.RenderPlayer ?? self.World.LocalPlayer;
+			if (IsTraitDisabled || !(viewer != null && Info.ValidStances.HasStance(self.Owner.Stances[viewer])))
```

Comment:
Using `IsTraitDisabled || viewer == null || !Info.ValidStances...` looks simpler to me (I don't really mind this either, though).